### PR TITLE
Java: Add the `XREADGROUP` command

### PR DESCRIPTION
**Pull request by acarbonetto**
Dependent on https://github.com/aws/glide-for-redis/pull/1599/

Adds the [XREADGROUP](https://valkey.io/commands/xreadgroup/) command to Java

TODO: 

- [X] Fix IT tests for XREADGROUP
- [x] Fix IT test for failure/blocking cases

### CODE REVIEW

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 1]
```diff
@@ -3009,6 +3013,85 @@ public T xgroupDestroy(@NonNull String key, @NonNull String groupname) {
...
+     * @param keysAndIds A <code>Map</code> of keys and entry ids to read from. The <code>
+     *     Map</code> is composed of a stream's key and the id of the entry after which the stream
+     *     will be read. Use the special id of <code>{@literal ">"}</code> to receive only new messages.
+     * @param group The consumer group name.
+     * @param consumer The newly created consumer.
+     * @return Command Response - A <code>{@literal Map<String, Map<String[][]>>}</code> with stream
+     *      keys, to <code>Map</code> of stream-ids, to an array of pairings with format <code>[[field, entry], [field, entry], ...]<code>.
+     *      Returns null if the consumer group does not exist. Returns a Map with a value of null if the stream is empty.
+     */
+    public T xreadgroup(Map<String, String> keysAndIds, String group, String consumer) {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3065]: Add missing nonnulls

File: java/client/src/main/java/glide/api/models/commands/stream/StreamReadOptions.java [hunk 2]
```diff
@@ -7,15 +7,15 @@
...
+import lombok.experimental.SuperBuilder;
 
 /**
  * Optional arguments for {@link StreamBaseCommands#xread(Map, StreamReadOptions)}
  *
  * @see <a href="https://redis.io/commands/xread/">redis.io</a>
  */
-@Builder
-public final class StreamReadOptions {
+@SuperBuilder
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/stream/StreamReadOptions.java:17]: <3

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java [hunk 3]
```diff
@@ -128,7 +129,10 @@ public static <T> Map<String, T[][]> castMapOf2DArray(
             return null;
         }
         return mapOfArrays.entrySet().stream()
-                .collect(Collectors.toMap(Map.Entry::getKey, e -> castArrayofArrays(e.getValue(), clazz)));
+                .collect(
+                        HashMap::new,
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/utils/ArrayTransformUtils.java:133]: why this change? to accept nulls?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 4]
```diff
@@ -3488,6 +3489,203 @@ public void xgroupCreate_xgroupDestroy(BaseClient client) {
...
+                                                consumerName)
+                                        .get());
+        assertInstanceOf(RequestException.class, executionException.getCause());
+
+        executionException =
+                assertThrows(
+                        ExecutionException.class,
+                        () ->
+                                client
+                                        .xreadgroup(
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3619]: what is the difference vs previous check?

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 5]
```diff
@@ -753,6 +771,18 @@ private static Object[] streamCommands(BaseTransaction<?> transaction) {
             1L, // xtrim(streamKey1, new MinId(true, "0-2"))
             OK, // xgroupCreate(streamKey1, groupName1, "0-0")
             OK, // xgroupCreate(streamKey1, groupName1, "0-0", options)
+            true, // xgroupCreateConsumer(streamKey1, groupName1, consumer1)
+            Map.of(
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:775]: nit: you can store this as a variable if it ease reading/debugging

File: java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java [hunk 6]
```diff
@@ -81,6 +81,10 @@ public void transactions_with_group_of_commands(String testName, TransactionBuil
         Object[] expectedResult = builder.apply(transaction);
 
         Object[] results = clusterClient.exec(transaction).get();
+        if (testName.equals("Stream Commands")) {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java:84]: please delete

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 7]
```diff
@@ -372,4 +373,122 @@ CompletableFuture<String> xgroupCreate(
...
+     * @apiNote When in cluster mode, all keys in <code>keysAndIds</code> must map to the same hash
+     *     slot.
+     * @see <a href="https://valkey.io/commands/xreadgroup/">valkey.io</a> for details.
+     * @param keysAndIds A <code>Map</code> of keys and entry ids to read from. The <code>
+     *     Map</code> is composed of a stream's key and the id of the entry after which the stream
+     *     will be read. Use the special id of <code>{@literal ">"}</code> to receive only new messages.
+     * @param group The consumer group name.
+     * @param consumer The newly created consumer.
+     * @param options Options detailing how to read the stream {@link StreamReadGroupOptions}.
+     * @return A <code>{@literal Map<String, Map<String[][]>>}</code> with stream
```
Comment (GumpacG) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:463]:
```suggestion
     * @return A <code>{@literal Map<String, Map<String, Map<String[][]>>>}</code> with stream
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 8]
```diff
@@ -3009,6 +3013,85 @@ public T xgroupDestroy(@NonNull String key, @NonNull String groupname) {
...
+     * @apiNote When in cluster mode, all keys in <code>keysAndIds</code> must map to the same hash
+     *     slot.
+     * @see <a href="https://valkey.io/commands/xreadgroup/">valkey.io</a> for details.
+     * @param keysAndIds A <code>Map</code> of keys and entry ids to read from. The <code>
+     *     Map</code> is composed of a stream's key and the id of the entry after which the stream
+     *     will be read. Use the special id of <code>{@literal ">"}</code> to receive only new messages.
+     * @param group The consumer group name.
+     * @param consumer The newly created consumer.
+     * @param options Options detailing how to read the stream {@link StreamReadGroupOptions}.
+     * @return Command Response - A <code>{@literal Map<String, Map<String[][]>>}</code> with stream
```
Comment (GumpacG) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:3081]:
```suggestion
     * @return Command Response - A <code>{@literal Map<String, Map<String, Map<String[][]>>>}</code> with stream
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 9]
```diff
@@ -3488,6 +3489,203 @@ public void xgroupCreate_xgroupDestroy(BaseClient client) {
...
+
+        // Add two stream entries
+        String streamid_1 = client.xadd(key, Map.of("field1", "value1")).get();
+        assertNotNull(streamid_1);
+        String streamid_2 = client.xadd(key, Map.of("field2", "value2")).get();
+        assertNotNull(streamid_2);
+
+        // read the entire stream for the consumer and mark messages as pending
+        var result_1 = client.xreadgroup(Map.of(key, ">"), groupName, consumerName).get();
+        assertEquals(2, result_1.get(key).size());
```
Comment (GumpacG) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3533]: Can we instead check for the whole result object?

**Review by GumpacG — APPROVED** [review]

File: glide-core/src/client/value_conversion.rs [hunk 10]
```diff
@@ -852,17 +853,29 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
...
+        b"HEXISTS"
+        | b"HSETNX"
+        | b"EXPIRE"
+        | b"EXPIREAT"
+        | b"PEXPIRE"
+        | b"PEXPIREAT"
+        | b"SISMEMBER"
+        | b"PERSIST"
+        | b"SMOVE"
+        | b"RENAMENX"
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply glide-core/src/client/value_conversion.rs:873]: needs rebasing...

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java (hunk 3 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/utils/ArrayTransformUtils.java:133]: yes.  XREADGROUP returns maps will null values (gross!)

File: glide-core/src/client/value_conversion.rs [hunk 11]
```diff
@@ -791,6 +791,7 @@ fn convert_to_array_of_pairs(
     value_expected_return_type: Option<ExpectedReturnType>,
 ) -> RedisResult<Value> {
     match response {
+        Value::Nil => Ok(response),
```
Comment (acarbonetto) [inline glide-core/src/client/value_conversion.rs:794]: XREADGROUP can/will return null values in the map

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/utils/ArrayTransformUtils.java:133]: Do you have a test for that case?

Comment (acarbonetto) [reply java/client/src/main/java/glide/utils/ArrayTransformUtils.java:133]: I do

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 9 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3522]: sure

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 5 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:783]: might be a good idea... ;)

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 4 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3664]: the key is valid in this case... I need a comment

File: glide-core/src/request_type.rs [hunk 12]
```diff
@@ -192,6 +192,8 @@ pub enum RequestType {
     GeoSearch = 182,
     Watch = 183,
     UnWatch = 184,
+    XGroupCreateConsumer = 185,
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply glide-core/src/request_type.rs]: rebase... ignore this
